bpf/bpf_lxc.c

```diff
@@ -650,6 +650,13 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 		key.ip6.p4 = 0;
```

Comment:
> However, always passing packets to the stack breaks the host firewall policy enforcement for pod to node traffic, as at that point there's no route which redirects these packets back to the tunnel to preserve the security identity, and they get simply masqueraded and routed natively.
> 
> To prevent this issue, let's pass packets to the stack only if they are a reply with destination identity matching a remote node, as in that case they may need to be rev DNATted.

I'm a bit confused here. So passing packets to the stack means they are not sent through the tunnel and we lose the source identity. You propose to be more selective and only send reply packets to remote nodes to the stack. I understand that will limit the impact (less cases where we lose the source identity), but it won't completely fix the issue. Replies from a pod to a remote node will still lose their source identity. What am I missing?

Reply:
After thinking a bit more about this (and doing some tests), my understanding is that node to pod traffic would not be affected. AFAIU, there are two possible cases depending on the source address:

* CiliumInternalIP -> remote pod: the packet is redirected to `cilium_host` by the route covering the remote PodCIDR, processed by `cil_from_host`, forwarded through the tunnel and eventually to the remote pod. The reply would then be processed by `cil_from_container`, and passed to the stack as per the above changes (given that the destination identity is `remote_node`). At that point, it would be redirected to `cilium_host` (as the CiliumInternalIP belongs to the remote PodCIDR), forwarded through the tunnel (with the correct source identity, [configured by `cil_from_host`](https://github.com/giorio94/cilium/blob/mio%2Fhost-firewall-tunnel/bpf/bpf_host.c#L1134-L1135)), eventually reaching the origin node again. The source IP is also preserved, so the identity could be looked up on the destination node.
* InternalIP -> remote pod: the packet is redirected to `cilium_host` as before, source natted to the CiliumInternalIP address (third rule below) and then continue along the lines of the previous case. 
   ```
  Chain CILIUM_POST_nat (1 references)
  target     prot opt source               destination
  MASQUERADE  all  --  10.244.2.0/24       !10.244.2.0/24        /* cilium masquerade non-cluster */
  ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0            mark match 0xa00/0xe00 /* exclude proxy return traffic from masquerade */
  SNAT       all  -- !10.244.2.0/24       !10.244.2.0/24        /* cilium host->cluster masquerade */ to:10.244.2.118
  SNAT       all  --  127.0.0.1            0.0.0.0/0            /* cilium host->cluster from 127.0.0.1 masquerade */ to:10.244.2.118
  SNAT       all  --  0.0.0.0/0            0.0.0.0/0            mark match 0xf00/0xf00 ctstate DNAT /* hairpin traffic that originated from a local pod */ to:10.244.2.118
   ```

I guess we could be even more specific in limiting the packets passed to the stack, and do that only if the original packet didn't come from the tunnel (`!ct_state->from_tunnel`). My understanding, though, is that at the moment it gets set [only in the IPv4 path](https://github.com/giorio94/cilium/blob/mio%2Fhost-firewall-tunnel/bpf/bpf_lxc.c#L1971), and would require some additional changes to make it work for IPv6 as well, making this fix a little harder to backport (depending on if and to which stable branches we would backport it).

Reply:
> Assuming a connection node1 --VIP--> pod@node2 (VIP is node2IP).

I assume you are talking about the ExternalTrafficPolicy=Local case here.

> When the SYN+ACK packet reaches node1, we lost its source identity (i.e., pod@node2).

I wonder which identity would be even correct in this case, as from the PoW of node1, the destination identity would be remote_node (as the dst IP is that of the node), and not the pod identity. And it would be matched by a CCNP which allows egress traffic towards the node, not the pod. So it sounds reasonable to me that also the reply would have a source identity corresponding to the node.

My understanding is that the same would also apply in case of a service with ExternalTrafficPolicy=Cluster. A special case would then be `node1 --> VIP@node2 --> pod@node3`. The `node1->node2` path would be outside of the tunnel (with dst -- src for the reply -- identity remote_node), while `node2->node3` inside (given the SNATting performed on node2), (with dst -- src for the reply -- identity pod).

Reply:
> I assume you are talking about the ExternalTrafficPolicy=Local case here.

Yep.

> I wonder which identity would be even correct in this case, as from the PoW of node1, the destination identity would be remote_node (as the dst IP is that of the node), and not the pod identity. And it would be matched by a CCNP which allows egress traffic towards the node, not the pod. So it sounds reasonable to me that also the reply would have a source identity corresponding to the node.

Hm, maybe you're right. I didn't see things from that angle.